rustls/src/error.rs

```diff
@@ -66,6 +66,10 @@ pub enum Error {
     /// implementation.
     InvalidCertificate(CertificateError),
 
+    /// A provided certificate revocation list (CRL) was invalid.
+    // TODO(@cpu): provide more detail, maybe with a sub-enum, or by exposing webpki::Error?
+    InvalidCrl,
```

Comment:
This is one point I'd appreciate some guidance on - How do we want to handle the `webpki::Error`s returned by `CertRevocationList::from_der`.

We could:

* Follow the lead of `CertificateError` and make a sub-enum that has Rustls-specified versions of the possible error returns, and then impl a  `TryFrom<webpki::Error>` or something like `verify.rs`'s [`pki_error`](https://github.com/rustls/rustls/blob/5053fde25bb0fde5715a95601a62c8fae626600e/rustls/src/verify.rs#L653-L667) fn.
* Change `InvalidCrl` to something like `InvalidCrl(Arc<dyn StdError + Send + Sync>)` and wrap the `webpki::Error` in an arc as a generic error
* Leave it as-is and not provide more specificity.
* Something else?

Reply:
To what extent are CRL errors from webpki disjoint from other webpki errors? I do feel like we would definitely want to separate "error while parsing CRL" from "client certificate was revoked according to CRL" in our rustls error messaging.

In terms of timing, maybe we should also try to parse the `CertRevocationList`s once at `AllowAuthenticatedClient` construction just to draw out any errors from parsing?

Reply:
I need to dig in to where the "client certificate was revoked" error is getting bubbled up through Rustls. I have a feeling it's being converted to a Rustls internal error already. The error I added/flagged for discussion here is strictly from the "error while parsing CRL" codepath.

For the parsing errors, I think we can emit:
* UnsupportedCrlVersion
* UnsupportedDeltaCrl
* UnsupportedIndirectCrl
* UnsupportedRevocationReason
* InvalidCrlNumber
* InvalidSerialNumber
* SignatureAlgorithmMismatch
* UnknownIssuer
* MalformedExtensions
* BadDerTime
* BadDer (like almost all parts of webpki :sweat_smile:)

Of these I think probably the "Unsupported" ones are the most interesting to a user and most disjoint from existing errors. Most of the others (minus InvalidCrlNuber) are already used by webpki elsewhere.

On the path of checking a cert for revocation with a CRL since we front-load most of the error checking for the CRL itself I think we only emit:

* CertRevoked
* IssuerNotCrlSigner
* InvalidSignatureForPublicKey

Of these only CertRevoked and IssuerNotCrlSigner are new errors not previously possible for Rustls code to see.

Reply:
> In terms of timing, maybe we should also try to parse the CertRevocationLists once at AllowAuthenticatedClient construction just to draw out any errors from parsing?

That's a good idea :+1:

Reply:
Implemented up-front parsing in my recent push.

Reply:
Started on benchmarking: https://github.com/rustls/webpki/pull/94

Reply:
> I need to dig in to where the "client certificate was revoked" error is getting bubbled up through Rustls. I have a feeling it's being converted to/masked by a Rustls internal error already.

Traced this out. We were turning the `webpki::Error::CertRevoked` return from the verifier into a `InvalidCertificate(CertificateError::Other(...))` err. I had to add a match arm in `verify.rs`'s `pki_error` for `webpki::RevokedCert`. Conveniently there was already a `CertificateError::Revoked` on the Rustls side to map it to. I force pushed an update to this branch to include that change and in the "happy" revocation path we now get a sensible error bubbled up to the user:

Server side:
```
cargo run --package rustls-examples --bin tlsserver-mio -- 
  -p 4443 
  --certs ./pki/localhost/cert.pem 
  --key ./pki/localhost/key.pem 
  --auth ./pki/minica.pem
  --crl ./pki/demo.crl
  --crl ./pki/demo.2.crl
  http
```

Client side:
```
cargo run --bin tlsclient-mio -- 
  -p 4443 
  --http localhost 
  --cafile ./pki/minica.pem 
  --auth-key ./pki/localhost/key.pem 
  --auth-certs ./pki/localhost/cert.pem
```

```
TLS error: AlertReceived(CertificateRevoked)
Connection closed
```